feat(reporter): add `--hide-reporter-prefix` option

zirkelc opened the pull request:
Closes #7061 

Implements a `--reporter-hide-prefix` option to hide the project name from lifecycle outputs of running scripts.

I couldn't find the docs, so I assumed they will be created from the code somehow. Let me know if I should document it somewhere.

Please let me know if any changes regarding naming, tests, etc. is desired.

zkochan commented on cli/default-reporter/src/reporterForClient/reportLifecycleScripts.ts:
```diff
@@ -215,15 +216,16 @@ function highlightLastFolder (p: string) {
+  hideLifecyclePrefix: boolean,
   logObj: LifecycleLog
 ) {
   const prefix = formatLifecycleScriptPrefix(colorByPkg, cwd, logObj.wd, logObj.stage)
```
There is no need to format the prefix if it won't be printed.

zirkelc replied:
I left it in there, because the prefix is still needed for the output of the exit code and script name:

```ts
  if (typeof logObj['exitCode'] === 'number') {
    if (logObj['exitCode'] === 0) {
      return `${prefix}: Done`
    } else {
      return `${prefix}: Failed`
    }
  }
  if (logObj['script']) {
    return `${prefix}$ ${logObj['script'] as string}`
  }
```

Only the actual output from the script itself has no prefix.

zkochan commented on cli/default-reporter/src/index.ts:
```diff
@@ -75,7 +76,7 @@ export function initDefaultReporter (
   const subscription = output$
     .subscribe({
-      complete () {}, // eslint-disable-line:no-empty
+      complete () { }, // eslint-disable-line:no-empty
```
revert unrelated formatting changes please

zkochan reviewed (changes requested):
There are also tests in default reporter. Add there a new test too.

The docs are in a different repo. They can be updated after release.

Add a changeset. You can see in the CONTRIBUTING.md how to create changesets.

zirkelc commented:
I added a new test to `cli/default-reporter/test/reportingLifecycleScripts.ts`. I initially added no test to this file, because it only contains tests for `pnpm install` and the changes I've implemented affect the `pnpm run` command.

Let me know what you think!

zirkelc commented on cli/default-reporter/test/reportingLifecycleScripts.ts:
```diff
@@ -233,29 +233,50 @@ test('groups lifecycle output when append-only is used', (done) => {
-  const allOutputs = [] as string[]
+  // const allOutputs = [] as string[]
 
-  output$.pipe(take(11), map(normalizeNewline)).subscribe({
-    complete: () => {
-      expect(allOutputs.join(EOL)).toBe(`\
-${chalk.cyan('packages/foo')} ${PREINSTALL}$ node foo
-${chalk.cyan('packages/foo')} ${PREINSTALL}: foo 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30
-${chalk.cyan('packages/foo')} ${PREINSTALL}: Failed
-${chalk.cyan('packages/foo')} ${POSTINSTALL}$ node foo
-${chalk.cyan('packages/foo')} ${POSTINSTALL}: foo I
-${chalk.magenta('packages/bar')} ${POSTINSTALL}$ node bar
-${chalk.magenta('packages/bar')} ${POSTINSTALL}: bar I
-${chalk.cyan('packages/foo')} ${POSTINSTALL}: foo II
-${chalk.cyan('packages/foo')} ${POSTINSTALL}: foo III
-${chalk.blue('packages/qar')} ${INSTALL}$ node qar
-${chalk.blue('packages/qar')} ${INSTALL}: Done`)
-      done()
-    },
-    error: done,
-    next: (output: string) => {
-      allOutputs.push(output)
-    },
-  })
```
This test would never complete, because if the `expect(allOutputs.join(EOL))` fails, the succeeding `done()` would not be called because `expect` throws

zirkelc commented on cli/default-reporter/test/reportingLifecycleScripts.ts:
```diff
@@ -490,10 +511,11 @@ ${chalk.cyan('packages/foo')} ${POSTINSTALL}: Done`)
+      appendOnly: true,
       outputMaxWidth: 79,
       hideLifecyclePrefix: true,
     },
```
The `--reporter-hide-prefix` only works for `append-only`, so I changed this test

zirkelc commented:
I noticed that `append-only` reporter is automatically active on CI and the `--reporter-hide-prefix` makes only sense for this reporter (and is only implemented for this). Therefore, I changed the new test case to use `appendOnly` and `hideLifecyclePrefix`.

For an unknown reason, the existing test case `groups lifecycle output when append-only is used` fails on my machine because it shows that there is an additional space char:

![image](https://github.com/pnpm/pnpm/assets/950244/b84ffbce-68b2-4fe0-aecc-ace63f887cc6)

I couldn't determine the reason for this, maybe you could give me a hint where I should look for to fix this?

zirkelc commented:
Hi @zkochan I just saw the new release. However, the option name is wrong in the release notes. The release notes say `--hide-reporter-prefix`, but the option is actually `--reporter-hide-prefix`.

I think it's may fault because I messed it in the commit description - sorry about that!

Could you change the release note manually?
